### DISCUSSION TITLE
Selectively choose which parameters to offload

### DISCUSSION THREAD

**Discussion by taehyunzzz** [deepspeedai/DeepSpeed#3536] [Q&A] [answered]
Is there a way to select which parameters I want to offload?

For example I want to offload only the big model parameters like FFN to CPU memory and fetch them when needed,
whereas all the other layers such as convolution stays in the GPU for good.

Does the current deepspeed allow such implementation?
If not, has anybody seen a custom user-implementation of such thing?
Where should I mainly modify to achieve this?

**Comment by tjruwase** [comment, marked as answer]
Try `"stage3_param_persistence_threshold"` to ensure that parameters smaller than a threshold are not partitioned or offloaded. 

See [here](https://www.deepspeed.ai/docs/config-json/#zero-optimizations-for-fp16-training) for details.
<img width="645" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/4271600/ba524c74-909c-4a64-963d-5b6e73447ef1">

**Reply by taehyunzzz** [reply]
@tjruwase Thank you. I have another question regarding ZeRO-inference blog [post](https://www.deepspeed.ai/2022/09/09/zero-inference.html).
They mentioned pinning some parameters to the GPU memory, rather than allowing swap between GPU and CPU.
Does the current deepspeed API provide such parameter pinning functionality?

**Reply by ethansmith2000** [reply]
are there alternative methods, say, if i wanted to offload non-trainable parameters?

**Reply by tjruwase** [reply]
@ethansmith2000, there are no methods for offloading only non-trainable parameters.

**Comment by tjruwase** [comment]
Use the following:  https://deepspeed.readthedocs.io/en/latest/zero3.html#deepspeed.runtime.zero.config.DeepSpeedZeroConfig.model_persistence_threshold
